static/js/silhouette.js

```diff
@@ -0,0 +1,121 @@
+'use strict'
+
+/**
+ * Calculate euclidean distance between two points.
+ * @function euclidean
+ * @param {number[]} x - coordinate of point x
+ * @param {number[]} y - coordinate of point y
+ * @return {number} euclidean distance between x and y
+ */
+function euclidean(x, y) {
+  // check x, y
+  if (arrEqual(x, y)) {
+    return 0;
+  }
+  let sum = 0;
+  for (let i = 0; i < x.length; i++) {
+    sum += (x[i] - y[i]) ** 2;
+  }
+  return Math.sqrt(sum);
+}
+
+/**
+ * Return the pairwise distance matrix of each point in the input data array.
+ * @function pdist
+ * @param {number[]} x - the input data array
+ * @return {number[]} distance matrix of the given data
+ */
+//TODO: add metrics='euclidean'
+function pdist(x) {
+  let d = Array(x.length).fill().map(()=>Array(x.length).fill());
+  for (let i = 0; i < x.length; i++) {
+    for (let j = 0; j < x.length; j++) {
+      if (j === i) {
+        d[i][j] = 0;
+      } else if (j < i) {
+        d[i][j] = d[j][i];
+      } else {
+        d[i][j] = euclidean(x[i], x[j]);
+      }
+    }
+  }
+  return d;
+}
+
+
+/**
+ * Return the occurrence of each entry in the input data.
+ * @function bincount
+ * @param {number[]} x - the input data array
+ * @return {number[]} the occurrence of each entry in the input data
+ */
+function bincount(x) {
+  let res = Array(Math.max.apply(Math, x) + 1).fill(0);
```

Comment:
You haven't addressed my question about the 2nd `Math`. I thought it should be `null`?

Reply:
fixed. I found some useful [notes](https://stackoverflow.com/questions/21255138/how-does-the-math-max-apply-work) about Math.max.apply(Math/null/undefined, array).